src/ripple/basics/impl/Number.cpp

```diff
@@ -0,0 +1,622 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2022 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/basics/Number.h>
+#include <algorithm>
+#include <cassert>
+#include <numeric>
+#include <stdexcept>
+#include <type_traits>
+
+#ifdef _MSVC_LANG
+#include <boost/multiprecision/cpp_int.hpp>
+using uint128_t = boost::multiprecision::uint128_t;
+#else   // !defined(_MSVC_LANG)
+using uint128_t = __uint128_t;
+#endif  // !defined(_MSVC_LANG)
+
+namespace ripple {
+
+// guard
+
+class Number::guard
+{
+    std::uint64_t digits_;
+    std::uint8_t xbit_ : 1;
+    std::uint8_t sbit_ : 1;  // TODO :  get rid of
+
+public:
+    explicit guard() : digits_{0}, xbit_{0}, sbit_{0}
+    {
+    }
+
+    void
+    set_positive() noexcept;
```

Comment:
Keep for now.  We can easily remove it later since it is not part of the public API.  I struggled a bit to get `Guard` right, and may need this if bugs turn up in this area.